### PR TITLE
Fix various CMake and GCC Warnings

### PR DESCRIPTION
**Pull request by DavidFair**
**Description of work.**
Fixes various CMake and gcc-8 Warnings, which have been annoying me locally for a while

- Gives our download targets names on CMake
- Removes the old policy for Span and Eigen, since they aren't specifying `fvisibility` so hiding the flag from un-exported targets shouldn't apply
- Suppresses the invalid cast warnings on Python/C++ boundaries for GCC 8> but leaves this warning on for the rest of the project where this is valid (minus MantidPlot)
- Fixes some copied exceptions

**To test:**
- Code review
- Attempt to configure the project using GCC-7 (default)
- Attempt to configure and build PythonKernelModule to check there is no warnings with GCC-8
`CC=gcc-8 CXX=g++-8 cmake path/to/mantid/src -G Ninja && ninja PythonKernelModule`

There is no associated issue. - Part of the sanitizer maintenance series

This does not require release notes because internal only changes

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/test/CalculatePlaczekSelfScatteringTest.h [hunk 1]
```diff
@@ -156,7 +156,7 @@ class CalculatePlaczekSelfScatteringTest : public CxxTest::TestSuite {
     alg->setProperty("IncidentSpecta", IncidentSpecta);
     alg->setProperty("InputWorkspace", InputWorkspace);
     alg->setProperty("OutputWorkspace", "correction_ws");
-    TS_ASSERT_THROWS(alg->execute(), std::runtime_error)
+    TS_ASSERT_THROWS(alg->execute(), std::runtime_error &)
```
Comment (martyngigg) [inline Framework/Algorithms/test/CalculatePlaczekSelfScatteringTest.h]: Most places use `const exc_type &`. Can we use this here for consistency?

Comment (DavidFair) [reply Framework/Algorithms/test/CalculatePlaczekSelfScatteringTest.h]: I've gone through and added `const` to all of the changed locations for general consistency

File: Framework/PythonInterface/mantid/kernel/CMakeLists.txt [hunk 2]
```diff
@@ -131,6 +131,10 @@ add_library(PythonKernelModule
             ${PYTHON_INSTALL_FILES})
 set_python_properties(PythonKernelModule _kernel)
 
+# GCC 8 onwards needs to disable functional casting at the Python interface
+target_compile_options( PythonKernelModule PRIVATE
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/kernel/CMakeLists.txt]: As far as I can tell this affects the places where a function pointer with a default parameter is cast to a function pointer with no parameters. I think this only affects the `Material.cpp` file? Should we suppress it more locally with the macros in `WarningSuppressions.h` instead? Or the other way around this in `Material.cpp` would be to put in a local function that takes no arguments and delegates to `Material::cohScatterXSection` and the others. That way we wouldn't need to suppress the warning?

Am I missing something?

Comment (martyngigg) [reply Framework/PythonInterface/mantid/kernel/CMakeLists.txt]: @DavidFair Have you had a chance to look at this.

Comment (DavidFair) [reply Framework/PythonInterface/mantid/kernel/CMakeLists.txt]: Not yet, planning on resolving this today or tomorrow. Thanks for the nudge as I had forgotten about it

**Review by martyngigg — APPROVED** [review]

**Comment by NickDraper** [comment]
Test failure unrelated (Kafka) merging

**Pull request merged**